src/containers/ParticleContainerInterface.h

```diff
@@ -0,0 +1,135 @@
+/*
+ * ParticleContainer.h
+ *
+ *  Created on: 17 Jan 2018
+ *      Author: tchipevn
+ */
+
+#pragma once
+
+#include <array>
+
+#include "iterators/ParticleIteratorWrapper.h"
+
+namespace autopas {
+
+// consider multiple inheritance or delegation vor avoidane of virtual call to
+// Functor
+/**
+ * The ParticleContainer class stores particles in some object and provides
+ * methods to iterate over its particles.
+ * @tparam Particle Class for particles
+ * @tparam ParticleCell Class for the particle cells
+ */
+template <class Particle>
+class ParticleContainerInterface {
+ public:
+  ParticleContainerInterface() {}
+
+  /**
+   * destructor of ParticleContainerInterface
+   */
+  virtual ~ParticleContainerInterface() = default;
+
+  /**
+   * delete the copy constructor to prevent unwanted copies.
+   * No particle container should ever be copied.
+   * @param obj
+   */
+  ParticleContainerInterface(const ParticleContainerInterface &obj) = delete;
+
+  /**
+   * delete the copy assignment operator to prevent unwanted copies
+   * No particle container should ever be copied.
+   * @param other
+   * @return
+   */
+  ParticleContainerInterface &operator=(const ParticleContainerInterface &other) = delete;
+
+  /**
+   * adds a particle to the container
+   * @param p the particle to be added
+   */
+  virtual void addParticle(Particle &p) = 0;
+
+  /**
+   * adds a particle to the container that lies in the halo region of the
+   * container
+   * @param haloParticle particle to be added
+   */
+  virtual void addHaloParticle(Particle &haloParticle) = 0;
+
+  /**
+   * deletes all halo particles
+   */
+  virtual void deleteHaloParticles() = 0;
+
+  /**
+   * iterate over all particles using
+   * for(auto iter = container.begin(); iter.isValid(); ++iter)
+   * @return iterator to the first particle
+   * @todo implement IteratorBehavior
+   */
+  virtual ParticleIteratorWrapper<Particle> begin(IteratorBehavior behavior = IteratorBehavior::haloAndOwned) = 0;
+
+  /**
+   * iterate over all particles in a specified region
+   * for(auto iter = container.getRegionIterator(lowCorner,
+   * highCorner);iter.isValid();++iter)
+   * @param lowerCorner lower corner of the region
+   * @param higherCorner higher corner of the region
+   * @return iterator to iterate over all particles in a specific region
+   */
+  virtual ParticleIteratorWrapper<Particle> getRegionIterator(std::array<double, 3> lowerCorner,
+                                                              std::array<double, 3> higherCorner) = 0;
```

Comment:
this one does not care weather the given region is in the halo, owned or both areas, correct?

Reply:
correct, probably todo

Reply:
done